File: app/src/main/java/localhost/toolkit/widget/recyclerview/HeterogeneousListAdapter.java
package localhost.toolkit.widget.recyclerview;

import android.util.SparseIntArray;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.regex.Pattern;

public class HeterogeneousListAdapter extends ListAdapter<HeterogeneousRecyclerItem, RecyclerView.ViewHolder> implements Filterable {
    private Filter filter;
    private final HashMap<Class<?>, Integer> classToType;
    private final SparseIntArray typeToPos;

    public HeterogeneousListAdapter() {
        super(new DiffUtilItemCallback<>());
        setHasStableIds(true);
        classToType = new HashMap<>();
        typeToPos = new SparseIntArray();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return getItem(typeToPos.get(viewType)).onCreateViewHolder(LayoutInflater.from(parent.getContext()), parent);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int position) {
        getItem(position).onBindViewHolder(viewHolder);
    }

    @Override
    public void onViewRecycled(@NonNull RecyclerView.ViewHolder holder) {
        int pos = holder.getBindingAdapterPosition();
        if (pos != RecyclerView.NO_POSITION)
            getItem(pos).onViewRecycled(holder);
    }

    @Override
    public int getItemViewType(int position) {
        Integer type = classToType.get(getItem(position).getClass());
        return type == null ? super.getItemViewType(position) : type;
    }

    @Override
    public long getItemId(int position) {
        return getItem(position).hashCode();
    }

    @Override
    public void submitList(@Nullable List<HeterogeneousRecyclerItem> list) {
        super.submitList(list, () -> {
            filter = null;
            updateTypes();
        });
    }

    @Override
    public void submitList(@Nullable List<HeterogeneousRecyclerItem> list, @Nullable Runnable commitCallback) {
        super.submitList(list, () -> {
            filter = null;
            updateTypes();
            if (commitCallback != null)
                commitCallback.run();
        });
    }

    private void applyFilter(@Nullable List<HeterogeneousRecyclerItem> list) {
        super.submitList(list, this::updateTypes);
    }

    private void updateTypes() {
        typeToPos.clear();
        for (int i = 0; i < getItemCount(); i++) {
            if (!classToType.containsKey(getItem(i).getClass()))
                classToType.put(getItem(i).getClass(), classToType.size());
            Integer type = classToType.get(getItem(i).getClass());
            if (type != null && typeToPos.indexOfKey(type) < 0)
                typeToPos.put(type, i);
        }
    }

    public Filter getFilter() {
        if (filter == null)
            filter = new HeterogeneousFilter();
        return filter;
    }

    private class HeterogeneousFilter extends Filter {
        private final List<HeterogeneousRecyclerItem> originalItems = new ArrayList<>(getCurrentList());

        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            FilterResults results = new FilterResults();
            if (constraint == null || constraint.length() == 0) {
                results.values = new ArrayList<>(originalItems);
                results.count = originalItems.size();
            } else {
                StringBuilder sb = new StringBuilder();
                sb.append("^\\X*");
                for (String word : constraint.toString().toLowerCase().split("\\W"))
                    sb.append("(?=.*").append(word).append(")");
                sb.append("\\X*$");
                ArrayList<HeterogeneousRecyclerItem> newValues = new ArrayList<>();
                Pattern pattern = Pattern.compile(sb.toString());
                for (HeterogeneousRecyclerItem value : originalItems)
                    if (pattern.matcher(value.toString().toLowerCase()).matches())
                        newValues.add(value);
                results.values = newValues;
                results.count = newValues.size();
            }
            return results;
        }

        @Override
        @SuppressWarnings("unchecked")
        protected void publishResults(CharSequence constraint, FilterResults results) {
            applyFilter((List<HeterogeneousRecyclerItem>) results.values);
        }
    }

    private static class DiffUtilItemCallback<I extends HeterogeneousRecyclerItem<?, ?>> extends DiffUtil.ItemCallback<I> {
        @Override
        public boolean areItemsTheSame(@NonNull I oldItem, @NonNull I newItem) {
            return oldItem.hashCode() == newItem.hashCode();
        }

        @Override
        public boolean areContentsTheSame(@NonNull I oldItem, @NonNull I newItem) {
            return oldItem.equals(newItem);
        }
    }
}

File: app/src/main/java/localhost/toolkit/widget/recyclerview/HeterogeneousListAdapter.kt
package localhost.toolkit.widget.recyclerview

import android.util.SparseIntArray
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import java.util.*
import java.util.regex.Pattern

class HeterogeneousListAdapter : ListAdapter<AbstractItemAdapter<*, *>, RecyclerView.ViewHolder>(DiffUtilItemCallback()), Filterable {
    private var filter: Filter? = null
    private val classToType: HashMap<Class<*>, Int> = HashMap()
    private val typeToPos: SparseIntArray = SparseIntArray()

    init {
        setHasStableIds(true)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        ViewHolder(getItem(typeToPos[viewType]).onCreateViewBinding(LayoutInflater.from(parent.context), parent))

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) =
        getItem(position).internalOnBindViewHolder(viewHolder)

    override fun onViewRecycled(holder: RecyclerView.ViewHolder) =
        holder.bindingAdapterPosition.let {
            if (it != RecyclerView.NO_POSITION) getItem(it).internalOnViewRecycled(holder)
        }

    override fun getItemViewType(position: Int) =
        classToType[getItem(position).javaClass] ?: super.getItemViewType(position)

    override fun getItemId(position: Int) =
        getItem(position).hashCode().toLong()

    override fun submitList(list: List<AbstractItemAdapter<*, *>>?) =
        super.submitList(list) {
            filter = null
            updateTypes()
        }

    override fun submitList(list: List<AbstractItemAdapter<*, *>>?, commitCallback: Runnable?) =
        super.submitList(list) {
            filter = null
            updateTypes()
            commitCallback?.run()
        }

    private fun applyFilter(list: List<AbstractItemAdapter<*, *>>?) =
        super.submitList(list) { updateTypes() }

    private fun updateTypes() {
        typeToPos.clear()
        for (i in 0 until itemCount) {
            if (!classToType.containsKey(getItem(i).javaClass))
                classToType[getItem(i).javaClass] = classToType.size
            classToType[getItem(i).javaClass].let { type ->
                if (type != null && typeToPos.indexOfKey(type) < 0)
                    typeToPos.put(type, i)
            }
        }
    }

    override fun getFilter(): Filter {
        if (filter == null) filter = HeterogeneousFilter()
        return filter!!
    }

    private inner class HeterogeneousFilter : Filter() {
        private val originalItems: List<AbstractItemAdapter<*, *>> = ArrayList(currentList)
        override fun performFiltering(constraint: CharSequence?): FilterResults {
            val results = FilterResults()
            if (constraint == null || constraint.isEmpty()) {
                results.values = ArrayList(originalItems)
                results.count = originalItems.size
            } else {
                val pattern = Pattern.compile(StringBuilder("^\\X*").apply {
                    constraint.toString().lowercase(Locale.getDefault()).split("\\W").forEach {
                        append("(?=.*").append(it).append(")")
                    }
                    append("\\X*$")
                }.toString())
                val newValues = ArrayList<AbstractItemAdapter<*, *>>()
                originalItems.forEach {
                    if (pattern.matcher(it.toString().lowercase(Locale.getDefault())).matches())
                        newValues.add(it)
                }
                results.values = newValues
                results.count = newValues.size
            }
            return results
        }

        override fun publishResults(constraint: CharSequence, results: FilterResults) {
            @Suppress("UNCHECKED_CAST")
            applyFilter(results.values as List<AbstractItemAdapter<*, *>>?)
        }
    }

    private class DiffUtilItemCallback : DiffUtil.ItemCallback<AbstractItemAdapter<*, *>>() {
        override fun areItemsTheSame(oldItem: AbstractItemAdapter<*, *>, newItem: AbstractItemAdapter<*, *>): Boolean {
            return oldItem.hashCode() == newItem.hashCode()
        }

        override fun areContentsTheSame(oldItem: AbstractItemAdapter<*, *>, newItem: AbstractItemAdapter<*, *>): Boolean {
            return oldItem == newItem
        }
    }
}

class ViewHolder<B : ViewBinding>(val binding: B) : RecyclerView.ViewHolder(binding.root)